pkg/util/limiter/memory_limiter.go

```diff
@@ -0,0 +1,242 @@
+	if !cfg.Enabled {
+		return &NoOpMemoryLimiter{}, nil
+	}
+
+	if cfg.CheckInterval <= 0 {
+		return nil, errCheckIntervalOutOfRange
+	}
+	if cfg.MemoryLimitBytes == 0 {
+		return nil, errLimitOutOfRange
+	}
+	if cfg.MemoryLimitBytes == 0 {
+		return nil, errLimitOutOfRange
+	}
+
+	ml := &HeapMemoryLimiter{
+		memCheckInterval: cfg.CheckInterval,
+		memLimit:         cfg.MemoryLimitBytes,
+		resetLimit:       cfg.ResetBytes,
+		ticker:           time.NewTicker(cfg.CheckInterval),
+		readMemStatsFn:   runtime.ReadMemStats,
+		signalCh:         make(chan struct{}),
+		logger:           logger,
+		memPressure:      &atomic.Bool{},
+	}
+
+	ml.start()
+	return ml, nil
+}
+
+type HeapMemoryLimiter struct {
+	// The limit in bytes above which the system is considered to be under pressure.
+	memLimit uint64
+	// The limit above which the system is considered healthy.
+	resetLimit uint64
+
+	// Channel used to signalCh goroutines that they can resume.
+	signalCh chan struct{}
+
+	// Requests queue and mutex
+	requests      []int64
+	requestsMutex sync.Mutex
+
+	// Denotes whether there is memory pressure.
+	memPressure *atomic.Bool
+
+	// The interval to check memory stats.
+	memCheckInterval time.Duration
+	// Ticker for watching memory stats.
+	ticker *time.Ticker
+
+	// The function to read the mem values is set as a reference to help with
+	// testing different values.
+	readMemStatsFn func(m *runtime.MemStats)
+
+	logger log.Logger
+}
+
+// Helper function to add the limiter to context.
+func AddMemLimiterToContext(ctx context.Context, limiter MemoryLimiter) context.Context {
+	return context.WithValue(ctx, mlCtxKey, limiter)
+}
+
+// Helper function to get the limiter from context.
+func MemLimiterFromContextWithFallback(ctx context.Context) MemoryLimiter {
+	ml, ok := ctx.Value(mlCtxKey).(MemoryLimiter)
+	if !ok {
+		ml = &NoOpMemoryLimiter{}
+	}
+	return ml
+}
+
+// Helper function to add the request ID to context.
+func AddRequestIDToContext(ctx context.Context, requestID int64) context.Context {
+	return context.WithValue(ctx, mlRidCtxKey, requestID)
+}
+
+// Helper function to get the request ID from context.
+func RequestIDFromContextWithFallback(ctx context.Context) int64 {
+	requestID, ok := ctx.Value(mlRidCtxKey).(int64)
+	if !ok {
+		return 0
+	}
+	return requestID
+}
+
+// Adds a request to the queue.
+func (ml *HeapMemoryLimiter) AddRequest(id int64) {
+	ml.requestsMutex.Lock()
+	defer ml.requestsMutex.Unlock()
+	ml.requests = append(ml.requests, id)
+}
+
+// Removes a request from the queue..
+func (ml *HeapMemoryLimiter) RemoveRequest(id int64) {
+	ml.requestsMutex.Lock()
+	defer ml.requestsMutex.Unlock()
+	for i, v := range ml.requests {
+		if v == id {
+			ml.requests = append(ml.requests[:i], ml.requests[i+1:]...)
+			break
+		}
+	}
+}
+
+// // If memory pressure is detected, waits will block until the memory pressure is mitigated or the context is done.
+// ID of 0 is special and will never be allowed if under memory pressure.
```

Comment:
This doesn't really reflect what the code is doing ... 

`ml.isFirstRequest(id)` returns `false` if id is `0`, and this function returns immediately; memory pressure is not checked at all. I guess maybe I don't understand what "allow" means here. 

Also, why make `0` special? If caller knows they are going pass in `0`, why can't the caller just not make a call to `Wait`?

Reply:
We need a way to allow the first request in the queue to continue when streaming from ingesters/store-gateways. This way the requests will execute one by one and eventually will all finish.

However, we also don't want to pull more requests from the QFE when under memory pressure. That was the original intent of using 0 as a special case. I now changed the wait to accept a param `allowFirst` that tells if the first request must be allowed even under pressure.